GQL framework (#212)
neo4j/docs-status-codes: modules/ROOT/pages/errors/index.adoc

:description: This page describes the structure of the error objects, the error codes, and how to interpret them.
[[neo4j-errors]]
= Server errors

> The Neo4j DBMS returns a server error to indicate that the outcome of a Cypher query or command execution is unsuccessful.

Reviewer: What is here "Neo4j DBMS" is elsewhere in the text "server", sometime "database", or just "Neo4j". It does not feel like the difference are alway deliberate. Maybe it is possible to have a consistent term for the thing the returns the error. I do not have a immediate proposal, unfortunately.

Author: Neo4j and Neo4j DBMS are the same thing. Neo4j DBMS or Neo4j can comprise one or more servers, which contain databases.

Author: I'll remove DBMS and leave only Neo4j.

The driver receives these erors and sends them to the Neo4j tools (e.g. Browser, Bloom, Cypher Shell) or the user application, which display them to the user.

Starting from version 5.26, the Neo4j error codes have an additional GQL-status object along with the Neo4j exception that provides information about the status of a Cypher query or command execution in compliance with the link:https://www.iso.org/standard/76120.html[ISO/IEC 39075:2024(en) - Information technology - Database languages - GQL Standard].

> This additional GQL-status object is also displayed in the query log from Neo4j 5.25.

Reviewer: "... from Neo4j 5.25"

Is here a word missing? Like "... from Neo4j 5.25 onwards" or "... starting from Neo4j 5.25" or such?

|===

For more information, see the xref:errors/all-errors.adoc[List of Neo4j error codes].

[[error-types]]
== Types of server errors

The fact that an error is returned by the server does not always mean that it is a fatal error.
Status codes can also indicate transient problems that may go away if you retry the request.
The server error group determines the effect on the transaction.

.Server error types
[options="header", cols="<1m,<2,<1"]
|===

| Type
| Description
| Effect on the transaction

| ClientError
| These errors are caused by the client (user input or user application) and are usually related to the request itself.
Changing the request might yield a successful outcome.
 Neo4j codes have the prefix `Neo.ClientError`, while the GQLSTATUS codes have `ErrorClassification` of type `CLIENT_ERROR`.
| Rollback

| TransientError
| These errors are detected by the server and are usually related to some kind of database unavailability, such as limits reached, out-of-memory, timeouts, etc.
The error can be temporary, therefore retrying later might yield a successful outcome.

> Neo4j codes have the prefix Neo.`TransientError`, while the GQLSTATUS codes have `ErrorClassification` of type `TRANSIENT_ERROR`.

Reviewer: ```suggestion
Neo4j codes have the prefix `Neo.TransientError`, while the GQLSTATUS codes have `ErrorClassification` of type `TRANSIENT_ERROR`.
```

| Rollback

| DatabaseError
| These errors are caused by the database and are usually related to the database state and mean that the database failed to service the request.
Neo4j codes have the prefix `Neo.DatabaseError`, while the GQLSTATUS codes have `ErrorClassification` of type `DATABASE_ERROR`.
| Rollback

|===

[[error-internals]]
== Error internals

Neo4j supports server errors in the form of Java exceptions.
Most of these implement the `HasStatus` interface, which means they have a status code in addition to the exception message.

On the server side, an exception contains normal Java constructors and methods like `getMessage()`,  `getCause()`, etc., and additionally the `status()` method from the `HasStatus` API, which returns the status code.

The exceptions also get new compulsary fields for `gqlStatus`, `statusDescription`,  `diagnosticRecord`, and an optional field for `cause`.

> The cause field in turn has its own GQLSTATUS, statusdescription, diagnostic record, and message. +

Reviewer: ```suggestion
The cause field in turn has its own GQLSTATUS, status description, diagnostic record, and message. +
```

All of these fields construct the GQLSTATUS object, which is sent to the driver as part of the Failure Bolt message.
Exactly how this looks, depends on the combination of driver and server versions.
See <<server-driver-compatibility, Server-driver version compatibility>> for more information.

On the driver side, the Neo4jException is extended with the corresponding methods as on the server side.
The driver receives the Failure Bolt message and extracts the status code and the error message.
Then, it constructs an exception with the status code, error message, and other relevant information, and sends it to the client.

// Starting from 2025.01, the diagnostic record also contains a `_status_parameters` field, which represens a map that contains all variable parts of the status description, such as labels, database names, Cypher clauses, etc.
// This field is used to provide more detailed information about the error.

== Query logging

Since the query log is server-side and DBMS wide, multiple clients connected to the same DBMS write to the same query log.
As the clients can have separate driver versions, they may have different error framework formats.

> In Neo4j 5.25, the default JSON template for the query log is updated to include `errorInfo` entry.

Reviewer: ```suggestion
In Neo4j 5.25, the default JSON template for the query log is updated to include an `errorInfo` entry.
```

!===
|===

A GQL-status error object can also contain an optional wrapped GQL-status object.
This wrapped object is referred to as a `cause` and used to provide additional, more specific diagnostic information.

[[gqlstatus-neo4j-defined-codes]]
== Classes of GQLSTATUS error codes

The GQLSTATUS codes are divided into classes and subclasses.
The class code is a 2-character string that indicates the general condition of the status, such as connection exception, data exception, etc.
The subclass code is a 3-character string that provides more detailed information about the condition.
The Neo4j-define subclasses start with N, followed by a 2-digit number, while the ones that do not start with N are standard-defined subclasses.

The following table lists the GQLSTATUS classes and their meanings:

.GQLSTATUS code classes
[frame="topbot", stripes=odd, grid="cols", cols="<1s,<4"]
|===
|Class
|*Condition*

| 08
| connection exception
| 22
| data exception
| 25
| invalid transaction state
| 40
| transaction rollback
| 42
| syntax error or access rule violation
| 50
| general processing exception
| 51
| system configuration or operation exception
| 52
| procedure exception
|===

> [[-error-object]]

Reviewer 2: ```suggestion
[[error-object]]
```

Starting from 5.25, the exceptions also get new compulsary fields for `gqlStatus`, `statusDescription`,  `diagnosticRecord`, and an optional field for `cause`.
The cause field is a wrapped underlying Neo4j exception, which in turn has its own GQLSTATUS, statusdescription, diagnostic record, and message. +
The `getMessage()` method is kept as Java exceptions inherently have this method.
And a new classification field is added to cover the division of client errors, transient errors, and database errors, which today is part of the Neo4j code.
All of these fields construct the GQLSTATUS object, which is sent to the driver as part of the Failure Bolt message.
Exactly how this looks, depends on the combination of driver and server versions.
See <<server-driver-compatibility, Server-driver version compatibility>> for more information.

On the driver side, the Neo4jException is extended with the corresponding methods as on the server side.
The driver receives the Failure Bolt message and extracts the status code and the error message.

> Then, it constructs an error object with the status code, error message, and other relevant information, and sends it to the client.

Reviewer 3: ```suggestion
Then, it constructs an exception with the status code, error message, and other relevant information, and sends it to the client.
```

:description: This page describes the structure of the error objects, the error codes, and how to interpret them.
[[neo4j-errors]]
= Server errors

The Neo4j DBMS returns a server error to indicate that the outcome of a Cypher query or command execution is unsuccessful.
The driver receives these errors and sends them to the client, which displays them to the user.

Starting from version 5.25, the Neo4j error codes have an additional GQL-status object that provides information about the status of a Cypher query or command execution in compliance with the link:https://www.iso.org/standard/76120.html[ISO/IEC 39075:2024(en) - Information technology - Database languages - GQL Standard].

This page describes the GQL-status object and the Neo4j error frameworks, their structure, the objects they provide for errors, and how to interpret them.
It also explains the error internals and the server-driver compatibility for both the GQLSTATUS and legacy errors.

[role=label--version-5.25]
[[gqlstatus-error-object]]
== GQL-status error object

> In the GQL-status object framework, errors are an implementation-defined subset of the GQL-status objects that cover exceptions.

Author: To change also for notifications.

In GQL, the execution of a query from user to server always has an outcome, called the _execution outcome_.

> The execution outcome is a list of GQL-status objects ordered by the following precedence rules:

Author: If we link this to the notifications, we also need to link the notifications to this page. Then, the question is, do we need this section in both places, or should we move it to another page that covers both?

> The execution outcome is a list of GQL-status objects ordered by the following precedence rules:

Author: Make it error-specific. Also change it for notifications.

. Every exception that results in a transaction rollback takes precedence over other exceptions.
. Every exception takes precedence over any completion condition.

The GQL-status object also includes the Neo4j-specific information, such as the severity level and the error classification.

Each GQL-status object consists of the following fields:

.GQLSTATUS error object
[cols="<1s,<4"]
|===
|GQLSTATUS code
a| A 5-character string that is the concatenation of a 2-character class code followed by a 3-character subclass code.
|StatusDescription

> a| A human-readable description of the GQLSTATUS, which consists of a condition, a subcondition, and a description.

Author: How about:
```suggestion
a| A human-readable description of the GQLSTATUS, which consists of a condition, a subcondition, and an optional additional text about the condition.
```

Starting from 5.25, the exceptions also get new compulsary fields for `gqlStatus` and `statusDescription`, and optional fields for the `diagnosticRecord` and `cause`.
The cause field is a wrapped underlying Neo4j exception, which in turn has its own GQLSTATUS, statusdescription, diagnostic record and message. +
The `getMessage()` method is kept as Java exceptions inherently have this method.
And a new classification field is added to cover the division of client errors, transient errors and database errors, which today is part of the Neo4j status code.
All of these fields construct the GQLSTATUS object, which is sent to the driver as part of the Failure Bolt message.
Exactly how this looks, depends on the combination of driver and server versions.
See <<server-driver-compatibility, Server-driver version compatibility>> for more information.

On the driver side, the Neo4jException is extended with the corresponding methods as on the server side.
The driver receives the failure message and extracts the status code and the error message.

> Then, it constructs an error object with the status code, error message, and other relevant information, and sends it to the client.

Author: Ask @bigmontz whether this sentence is correct.

> Then, it constructs an error object with the status code, error message, and other relevant information, and sends it to the client.

Reviewer 3: I think it is pretty much it. I don't know if it worth to mention that Neo4jException can also be generated on the driver side when drivers can't establish connection to the database, those errors was previously mapped as `ServiceUnavailable` and `SessionExpired` as neo4j code and they received status `50N42` for now since we don't have status code for them (and I don't know if we will have it).

In Neo4j 5.25, the default JSON template for the query log is updated to include `errorInfo` entry.
This entry contains `GQLSTATUS`, `statusDescription`, `classification`, `position` (if applicable), and `cause` (if applicable) with the same entries.

> The `failureReason` entry is deprecated from 5.26.

Reviewer 4: I cannot really remember if we did or not did deprecate this in the end?
